rasa/graph_components/providers/nlu_training_data_provider.py

```diff
@@ -0,0 +1,84 @@
+import os
+from rasa.engine.graph import GraphComponent, ExecutionContext
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.shared.importers.importer import TrainingDataImporter
+from rasa.shared.nlu.training_data.training_data import (
+    TrainingData,
+    DEFAULT_TRAINING_DATA_OUTPUT_PATH,
+)
+from rasa.shared.nlu.training_data.loading import load_data
+
+
+class NLUTrainingDataProvider(GraphComponent):
+    """Provides NLU training data during training and inference time."""
+
+    def __init__(
+        self,
+        model_storage: ModelStorage,
+        resource: Resource,
+        training_data: Optional[TrainingData] = None,
+    ) -> None:
+        """Creates NLU training data provider."""
+        self._model_storage = model_storage
+        self._resource = resource
+        self._training_data = training_data
+
+    @classmethod
+    def get_default_config(cls) -> Dict[Text, Any]:
+        """Get Default class configuration"""
+        return {"language": "en", "persist": False}
+
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> NLUTrainingDataProvider:
+        """Creates component (see parent class for full docstring)."""
+        return cls(model_storage, resource)
+
+    @classmethod
+    def load(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+        **kwargs: Any,
+    ) -> NLUTrainingDataProvider:
+        """Creates provider using a persisted data."""
+        with model_storage.read_from(resource) as resource_directory:
+            training_data = load_data(
+                resource_name=os.path.join(
+                    str(resource_directory), DEFAULT_TRAINING_DATA_OUTPUT_PATH
+                ),
+                language=config["language"],
+            )
+        return cls(model_storage, resource, training_data)
+
+    def _persist(self, training_data: TrainingData) -> None:
+        """Persists NLU training data to model storage."""
+        with self._model_storage.write_to(self._resource) as resource_directory:
+            training_data.persist(
+                dir_name=str(resource_directory),
+                filename=DEFAULT_TRAINING_DATA_OUTPUT_PATH,
+            )
+
+    def provide(
+        self, config: Dict[Text, Any], importer: Optional[TrainingDataImporter] = None,
+    ) -> TrainingData:
+        """Provides nlu training data during training."""
+        if config == {}:  # use default config if config is empty
+            config = self.get_default_config()
+        if importer is None:
+            training_data = self._training_data
+        else:
+            training_data = importer.get_nlu_data(language=config["language"])
```

Comment:
I wanted `self._training_data` to be accessible, so I made `importer` optional, and return `self._training_data` when no importer is passed. Not sure if this is the desired behaviour or if I should do something different.

Reply:
I think the other components (will) might always pass on an importer. But if we constructed this instance (self) via the `load` method, then there will be `self._training_data` that can be passed on. And if we didn't create this instance via `load`, then `self._training_data` will be `None`. So you can use the information whether `self._training_data` is None to decide whether the importer has to be used.

Downside of that approach is that if the data on disk changes, then a training data provider loaded from disk won't notice those differences (it will load the version of training data persisted elsewhere and not load the new data).

.... @joejuzl ?

Reply:
As we are not going to load the data, I don't think we need to even have `self._training_data`.